qiskit_machine_learning/algorithms/objective_functions.py

```diff
@@ -0,0 +1,158 @@
+
+    @abstractmethod
+    def objective(self, weights: np.ndarray) -> float:
+        """Computes the value of this objective function given weights.
+
+        Args:
+            weights: an array of weights to be used in the objective function.
+
+        Returns:
+            Value of the function.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def gradient(self, weights: np.ndarray) -> np.ndarray:
+        """Computes gradients of this objective function given weights.
+
+        Args:
+            weights: an array of weights to be used in the objective function.
+
+        Returns:
+            Gradients of the function.
+        """
+        raise NotImplementedError
+
+    def _neural_network_forward(self, weights: np.ndarray) -> Union[np.ndarray, SparseArray]:
+        """
+        Computes and caches the results of the forward pass. Cached values may be re-used in
+        gradient computation.
+
+        Args:
+            weights: an array of weights to be used in the forward pass.
+
+        Returns:
+            The result of the neural network.
+        """
+        data_id = str(id(weights)) + str(id(self._X))
+        if data_id != self._last_forward_id:
+            self._last_forward = self._neural_network.forward(self._X, weights)
+            self._last_forward_id = data_id
+        return self._last_forward
+
+
+class BinaryObjectiveFunction(ObjectiveFunction):
+    """An objective function for binary representation of the output,
+    e.g. classes of ``-1`` and ``+1``."""
+
+    def objective(self, weights: np.ndarray) -> float:
+        predict = self._neural_network_forward(weights)
+        target = np.array(self._y).reshape(predict.shape)
+        value = np.sum(self._loss(predict, target))
+        return value
+
+    def gradient(self, weights: np.ndarray) -> np.ndarray:
+        output = self._neural_network_forward(weights)
+        _, weights_grad = self._neural_network.backward(self._X, weights)
+
+        grad = np.zeros((1, self._neural_network.num_weights))
+        for i in range(len(self._X)):
+            grad += self._loss.gradient(output[i][0], self._y[i]) * weights_grad[i]
+
+        return grad
+
+
+class MultiClassObjectiveFunction(ObjectiveFunction):
+    """
+    An objective function for multiclass representation of the output,
+    e.g. classes of ``0``, ``1``, ``2``, etc.
+    """
+
+    def objective(self, weights: np.ndarray) -> float:
+        val = 0.0
+        probs = self._neural_network_forward(weights)
+        for i in range(len(self._X)):
+            for y_predict, prob in enumerate(probs[i]):
+                val += prob * self._loss(y_predict, self._y[i])
+        return val
+
+    def gradient(self, weights: np.ndarray) -> np.ndarray:
+        num_classes = self._neural_network.output_shape[0]
+        grad = np.zeros((1, self._neural_network.num_weights))
+        # TODO: do batch eval instead of zip()
```

Comment:
Was it intended to address the TODO in the current PR?

Reply:
As above, there's #63.